spec/inertia/conditional_sharing_spec.rb

```diff
@@ -1,28 +1,27 @@
+# Specs as documentation. Per-action shared data isn't explicity supported,
+# but it can be done by referencing the action name in an inertia_share block.
 RSpec.describe "conditionally shared data in a controller", type: :request do
-  context "when there is conditional data inside inertia_share" do
-    it "does not leak data between requests" do
-      get conditional_share_index_path, headers: {'X-Inertia' => true}
-      expect(JSON.parse(response.body)['props'].deep_symbolize_keys).to eq({
-                                                                             index_only_prop: 1,
-                                                                             normal_shared_prop: 1,
-                                                                           })
-
-      # NOTE: we actually have to run the show action twice since the new implementation
-      # sets up a before_action within a before_action to share the data.
-      # In effect, that means that the shared data isn't rendered until the second time the action is run.
-      get conditional_share_show_path, headers: {'X-Inertia' => true}
+  context "when there is data inside inertia_share only applicable to a single action" do
+    it "does not leak the data between requests" do
       get conditional_share_show_path, headers: {'X-Inertia' => true}
       expect(JSON.parse(response.body)['props'].deep_symbolize_keys).to eq({
-                                                                             normal_shared_prop: 1,
-                                                                             show_only_prop: 1,
-                                                                             conditionally_shared_show_prop: 1,
-                                                                           })
+        normal_shared_prop: 1,
+        show_only_prop: 1,
+        conditionally_shared_show_prop: 1,
+      })
 
       get conditional_share_index_path, headers: {'X-Inertia' => true}
-      expect(JSON.parse(response.body)['props'].deep_symbolize_keys).to eq({
-                                                                             index_only_prop: 1,
-                                                                             normal_shared_prop: 1,
-                                                                           })
+      expect(JSON.parse(response.body)['props'].deep_symbolize_keys).not_to include({
+        conditionally_shared_show_prop: 1,
+      })
+    end
+  end
+
+  context "when there is conditional data shared via before_action" do
+    it "raises an error because it is frozen" do
+      expect {
+        get conditional_share_show_with_a_problem_path, headers: {'X-Inertia' => true}
+      }.to raise_error(FrozenError)
```

Comment:
Does anyone by chance know why Rails 7.0 and lower doesn't raise an error here while Rails 7.1 does?

My guess is that maybe there was a change in the way Rails loads the code in 7.1, and previous versions don't actually load/freeze the data in the controller before this line gets called?

Reply:
We use Rails 7.1 `config.action_dispatch.show_exceptions = :none` config option for the dummy app, which evaluates to just truthy in Rails < 7.1 and acts like the `:all` option:
https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#new-ways-to-handle-exceptions-in-controller-tests-integration-tests-and-system-tests

Reply:
Yea, that was part of it but there was a second subtler bug. Because the shared data is frozen the first time you _access_ the shared data, if the first request to a controller has one of these (contrived) conditionally shared data examples, then it won't raise an error.

So, in the test suite, the failure (to raise a FrozenArray exception) was intermittent, depending on which spec in that file ran first.

It's an edge case of an edge case, so I don't feel we need to prevent it; and I still like having a spec to document the frozen data behavior.

In any case, thanks for #140 , it gave me a headstart tonight once I got back to looking at this one!

Reply:
Yeah, good catch. What do you think about marking `inertia_share` and `inertia_config` as `protected`? This way, calling them from action will raise an error, which might be a clear signal to users that they're doing something wrong.

Reply:
Yea, that might be cleaner. I'm not super worried about it either way now. It'll raise the FrozenArray error the second time you try to call the conditional action, so the failure I was seeing was just one of those test suite not quite matching reality quirks.

The per-action use case isn't actually officially supported at the moment, so we're being helpfully defensive enough already, IMO.